File: cpp/chapter2/layout.cpp
/* Layout
 * ベルマンフォード法
 */

#include <iostream>

using namespace std;

#define MAX_ML 10000
#define MAX_MD 1000000
#define MAX_N 1000
#define INF 9999999

// 入力
int N, ML, MD;
int AL[MAX_ML], BL[MAX_ML], DL[MAX_ML];
int AD[MAX_MD], BD[MAX_MD], DD[MAX_MD];

int d[MAX_N]; // 最短距離
bool updated; // 更新されたか

void update(int& x, int y){
  if(x > y){
    x = y;
    updated = true;
  }
}

// ベルマンフォード法によりdを計算する
void bellmanford(){
  for(int k = 0; k <= N; k++){
    updated = false;
    // i + 1からiへのコスト0
    for(int i = 0; i + 1 < N; i++){
      if(d[i + 1] < INF) update(d[i], d[i + 1]);
    }
    // ALからBLへコストDL
    for(int i = 0; i < ML; i++){
      if(d[AL[i] - 1] < INF) update(d[BL[i] - 1], d[AL[i] - 1] + DL[i]);
    }
    // BDからADへコスト-DD
    for(int i = 0; i < MD; i++){
      if(d[BD[i] - 1] < INF) update(d[AD[i] - 1], d[BD[i] - 1] - DD[i]);
    }
  }
}

int main(){
  cin >> N >> ML >> MD;
  for(int i = 0; i < ML; i++){
    cin >> AL[i] >> BL[i] >> DL[i];
  }
  for(int i = 0; i < MD; i++){
    cin >> AD[i] >> BD[i] >> DD[i];
  }

  // 閉路チェック
  fill(d, d + N, 0);
  bellmanford();
  if(updated){
    cout << -1 << endl;
    return 0;
  }

  fill(d, d + N, INF);
  d[0] = 0;
  bellmanford();
  int res = d[N -1];
  if(res == INF) res = -2;
  cout << res << endl;

  return 0;
}

File: python/chapter2/layout.py
"""
牛がN頭いて、番号順に並んでいる。
仲のいい牛と仲の悪い牛がいて、一定の距離を保ちたいと思っている。
複数の牛が同じ場所に並ぶことも可能。
ML個の仲のいい牛の情報(AL, BL, DL)とMD個の仲の悪い牛の情報(AD, BD, DD)が与えられる。
仲のいい牛は最大距離がDLとなるように、仲の悪い牛は最小距離がDDとなるように並べる。
この制約をみたす並び方のうち、1番の牛とN番の牛の最大距離を求める。
並び方が存在しない場合は-1を、いくつでも作れる場合は-2を出力する。

4
2
1 3 10
2 4 20
1
2 3 3
=>27
"""
INF = 99999999
n = int(input())
ml = int(input())
ML = []
for _ in range(ml):
    ML.append(list(map(int, input().split())))
md = int(input())
MD = []
for _ in range(md):
    MD.append(list(map(int, input().split())))


def bellmanford():
    for k in range(n):
        updated = False
        for i in range(n-1):
           if d[i+1] < INF:
               if d[i] > d[i+1]:
                   d[i] = d[i+1]
                   updated = True
        for al, bl, dl in ML:
            if d[al - 1] < INF:
                if d[bl-1] > d[al-1] + dl:
                    d[bl-1] = d[al-1] + dl
                    updated = True
        for ad, bd, dd in MD:
            if d[bd - 1] < INF:
                if d[ad-1] > d[bd-1] - dd:
                    d[ad-1] = d[bd-1] - dd
                    updated = True
    return updated

d = [0 for i in range(n)]
updated = bellmanford()
if updated:
    print(-1)
    exit()

d = [INF for i in range(n)]
d[0] = 0
bellmanford()
ans = d[n-1]

if ans == INF:
    ans = -2

print(ans)
